New merge algorithm (#2934)
mendix/docs: content/refguide/new-merge-algorithm.md

> ---

Author: The merge algorithm applies to all documents.

title: "New Merge Algorithm with Fine-Grained Conflict Resolution"
parent: "version-control"
tags: ["merge", "algorithm", "conflict", "resolution"]
---

## 1 Introduction
    
The new merge algorithm with fine-grained conflict resolution is used when you update your project or merge changes in it. The new algorithm has the following features: 
    
* **Fine-grained conflict resolution** – When there are conflicting changes in a document, you do not have to choose between whole documents: resolving a conflict using your change or using their change. Instead, you can resolve conflicts at the level of individual elements, such as widgets, entities, attributes, and microflow actions. Also, all non-conflicting changes from both sides are accepted automatically.
    

> * **No conflicts on parallel changes to lists of widgets** – When two developers make changes to the same list of widgets, there is no conflict, the changes are combined. However, if the changes are too close, a **list order conflict** is reported that reminds the developer who is merging the changes to decide on the final order of the widgets in the list. 

Reviewer: Depending on whether this is just for pages or can be applied to other documents - in which case the current name would be fine.

> * **No conflicts on parallel changes to lists of widgets** – When two developers make changes to the same list of widgets, there is no conflict, the changes are combined. However, if the changes are too close, a **list order conflict** is reported that reminds the developer who is merging the changes to decide on the final order of the widgets in the list. 

Author: It is about widgets inside the same container, e.g. data view or the page itself (for top-level widgets). 

And it also apply to other things like pages, e.g. the list of attributes in an entity or the values in an enumeration.

> * List order conflict – this option lists close changes to the same element and allows you to combine these changes. For example, two developers insert a widget in the same data container. These changes are too close and the merge algorithm cannot decide in what order to put these widgets, so there will be the so-called *list order conflict* that will allow you to place the widgets on the page in the right order. 

Author: List order conflicts are not the new feature here. The new feature is that changes to a list of widgets in some container do not result in real conflicts anymore. There may be a list order conflict if the changes are too close. So, can you rewrite this to be closer to what I originally wrote?

> * **Accepting one of the changes automatically in certain cases** – There are some cases where the new algorithm does not give a conflict. For example, one developer has moved an entity and the other has deleted it. The entity will be deleted in this case.

Author: Maybe we should just get rid of this third bullet. It is so small when compared to the other two.

> * **Accepting one of the changes automatically in certain cases** – There are some cases where the new algorithm does not give a conflict. For example, one developer has moved an entity and the other has deleted it. The entity will be deleted in this case.

Reviewer: I think it is important to keep this - it is not what I would expect to happen. If I move something but my colleague deletes it I would not expect it to disappear as I have moved it for a reason. If I didn't want it, I would have deleted it as well.

If we remove this, I think that we might get a lot of support calls saying that widgets are disappearing.

> * **Accepting one of the changes automatically in certain cases** – There are some cases where the new algorithm does not give a conflict. For example, one developer has moved an entity and the other has deleted it. The entity will be deleted in this case.

Author: It's only about moving an entity or microflow action to different coordinates (a meaningless change) and another person deleting the entity or microflow action. In that case we no longer raise a conflict because the first person's change is not important to keep. It does not apply to anything else. Not to widgets. And I think this third bullet will only confuse people.

> * **Accepting one of the changes automatically in certain cases** – There are some cases where the new algorithm does not give a conflict. For example, one developer has moved an entity and the other has deleted it. The entity will be deleted in this case.

Reviewer: @ironthorn - well it confused me, so I agree. Remove it.

2. The so-called **list order conflict**. Both people added widgets to the bottom layout grid. The merge algorithm cannot guess the right order of two new widgets and it reports the list order conflict. This is a reminder for the developer who is doing the merge to look at the final list. 

    ![New algorithm conflicts](attachments/new-merge-algorithm/new-merge-algorithm-conflicts.png)

To start the resolution process, click the **Resolve...** button. The page is opened in a special mode with an orange tab:

![Document with orange tab](attachments/new-merge-algorithm/new-merge-algorithm-orange-tab.png)

The following non-conflicting changes have already been applied to the page:

* The Mendix logo is added above the text *Home* (main line)
* The subtitle is deleted (main line)
* The text widget is added to the bottom layout grid (main line)
* A data grid is added to the bottom layout grid (branch line)

#### 3.3.1 Resolving the First Conflict

For the first conflict, you can inspect changes and decide which version to apply. Select one of the three lines that represent the conflict and choose **Resolve using Mine** or **Resolve using Theirs**. 

> ![Conflict resolution mode](attachments/new-merge-algorithm/new-merge-algorithm-resolve-mode.png)

Author: Yes! I am going to bring those names and buttons in line. And maybe even come up with much better names ;) Made a story for that.

You will see the document update immediately after you click the button. If you are not satisfied with your choice, you can use undo to go back and try another option. 

{{% alert type="info" %}}
To use keyboard shortcuts <kbd>Ctrl</kbd>+<kbd>Z</kbd> and <kbd>Ctrl</kbd>+<kbd>Y</kbd> to undo your choice, click the document to focus it first.
{{% /alert %}}

There is a third option to deal with a conflict: **Mark as Resolved**. This means that you do not choose any side to resolve the conflict and will keep things the way they were in the original.

Once you have chosen one of the three options to resolve the conflict, green checkmarks will appear to indicate that this conflict has been dealt with.

#### 3.3.2 Resolving the Second Conflict

The second conflict is a list order conflict. It is a reminder to take a look at the order of the widgets in the 
layout grid. You can arrange the widgets in the desired order in the page editor and then choose **Mark as Resolved** for the list order conflict.

> You can also decide to delete one of the widgets or add a new one. The page is fully editable while resolving conflicts. 

Reviewer: ```suggestion
You can also decide to delete one of the widgets or add a new one. The document is fully editable while resolving conflicts. 
```

> Some changes will make it impossible to resolve conflicts using **mine** or **theirs**. For example, if you had not resolved the first conflict yet and you delete the *Home* text widget, you cannot resolve the first conflict anymore, because the widget is simply not there. At that point, you can only mark the conflict as resolved:

Reviewer: ```suggestion
Some changes will make it impossible to resolve conflicts using **mine** or **theirs**. For example, if you have not resolved the first conflict yet and you delete the *Home* text widget, you cannot resolve the first conflict any more, because the widget is simply not there. At that point, you can only mark the conflict as resolved:
```

![Conflict cannot be resolved](attachments/new-merge-algorithm/new-merge-algorithm-cannot-resolve.PNG)

#### 3.3.3 Finishing Conflict Resolution

> Once all conflicts have been resolved, click the **Accept and Exit** button to finalize the results. The document will be saved in its merged form and the conflict for that document will be gone. The result is the document that contains changes from both sides and possibly some manual edits.

Reviewer: Is the **Accept and Exit** button disabled until all conflicts have been resolved? It doesn't appear to be greyed out in the image above.
If it isn't disabled, we need to explain what happens - e.g. can you save the unmerged version and come back to it later when you have discussed it with your colleague?

Author: It is not disabled but it will show a message that you cannot do it until you have resolved all conflicts. 

You cannot save the partially merged version and come back to it later. All active conflict resolutions are cancelled when closing the project / Studio Pro.
